keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,632 @@
+- On-premise deployment can’t horizontally scale available memory based on load https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-637715138 
+- Scaling resources is technically feasible but cost-prohibitive, swap provides flexibility at lower cost https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-553713502 
+
+This user story is addressed by scenario 2, and could benefit from 3.
+
+#### Local development and systems with fast storage
+
+Local development or single-node clusters and systems with fast storage may benefit from using available swap (e.g. NVMe swap partitions, one-node clusters).
+
+- Single node, local Kubernetes deployment on laptop https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-361748518 
+- Linux has optimizations for swap on SSD, allowing for performance boosts https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-589275277 
+
+This user story is addressed by scenarios 1 and 2, and could benefit from 3.
+
+#### Low footprint systems
+
+For example, edge devices with limited memory.
+
+- Edge compute systems/devices with small memory footprints (\<2Gi) https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-751398086 
+- Clusters with nodes \<4Gi memory https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-751404417 
+
+This user story is addressed by scenario 2, and could benefit from 3.
+
+#### Virtualization management overhead
+
+This would apply to virtualized Kubernetes workloads such as VMs launched by kubevirt.
+
+Every VM comes with a management related overhead which can sporadically be pretty significant (memory streaming, SRIOV attachment, gpu attachment, virtio-fs, …). Swap helps to not request much more memory to deal with short term worst-case scenarios.
+
+With virtualization, clusters are typically provisioned based on the workloads’ memory consumption, and any infrastructure container overhead is overcommitted. This overhead could be safely swapped out.
+
+- Required for live migration of VMs https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-754878431 
+
+This user story is addressed by scenario 2, and could benefit from 3.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Having swap available on a system reduces predictability. When swap is available to workloads, and is not accounted for on an individual workload-by-workload basis
+
+First, this risk is mitigated by preventing any workloads from using swap by default, even if it is enabled on a system. This will allow a cluster administrator to test swap utilization just at the system level without introducing unpredictability to workload resource utilization.
```

Comment:
I think this is being overly cautious for an opt-in initial version.  The admin can already restrict swap usage by _not_ enabling swap - and using existing mechanisms like node selectors / taints.  If we add a new option here, then we're just going to have to deprecate/remove it in future when we come up with the 'real' swap resource mechanism.

Reply:
I think this option (enable swap on node, but do not give it to workload) exists to make system tools on the node more reliable. So it is a valid scenario, that only addresses a very tiny set of user stories listed

Reply:
@anguslees IMO it's not realistic to try to phase in swap support faster than this. The flag already exists, and we can't remove it or change its behaviour in a single release :)

We have an existing CLI flag, we'll add a config with a default value to kubelet config. I don't know that we can avoid the feature flag, although maybe we could argue there's no point in having one if this can't actually be shut off at runtime. @johnbelamaric wdyt?